roofit/roofitcore/inc/RooFit/Detail/CodeSquashContext.h

```diff
@@ -0,0 +1,68 @@
+/*
+ * Project: RooFit
+ * Authors:
+ *   Garima Singh, CERN 2023
+ *   Jonas Rembser, CERN 2023
+ *
+ * Copyright (c) 2023, CERN
+ *
+ * Redistribution and use in source and binary forms,
+ * with or without modification, are permitted according to the terms
+ * listed in LICENSE (http://roofit.sourceforge.net/license.txt)
+ */
+
+#ifndef RooFit_Detail_CodeSquashContext_h
+#define RooFit_Detail_CodeSquashContext_h
+
+#include <string>
+#include <unordered_map>
+
+namespace RooFit {
+
+namespace Detail {
+
+/// @brief A class to maintain the context for squashing of RooFit models into code.
+class CodeSquashContext {
+   /// @brief Map of node names to their result strings.
+   std::unordered_map<std::string, std::string> nodeNames;
+   /// @brief Block of code that is placed before the rest of the function body.
+   std::string globalScope;
+
+public:
+   /// @brief Adds (or overwrites) the string representing the result of a node (aka "node result name").
+   /// @param key The name of the node to get the result name for.
+   /// @param value The new name to assign/overwrite.
+   inline void addResultName(std::string const &key, std::string value) { nodeNames[key] = value; }
+
+   /// @brief Gets the result name for the given node using the node name.
+   /// @param key Name of the node to get the result string for.
+   /// @return String representing the result of this node.
+   inline std::string getResultName(std::string const &key) { return nodeNames[key]; }
```

Comment:
Why did you call it `getResultName()`, `addResultName()` etc?

I think it's misleading. To me, this implies that it sets/returns a local variable name in the squashed code that stores the result. But if I understand correctly, this is not the case right? It actually deals with the *code* that evaluates to the result directly! So what does this have to do with the name of anything?

Why not just `getCode()`, and `setCode()`? Anyway that's just a suggestion, feel free to come up with a better name. But maybe something that doesn't have `Name` in it.

Reply:
Hmmm, not exactly. It deals with the name of the variable (it could also just be an expression) that the current node puts its result in. For example, a class implement a translates which outputs these statements:

```cpp
Int x = std::pow(y);
int z = x*y + 5;
```
Here, we want to "return" z from this node. That is why, we set the "name" of the "return" value of the function and not the code. This "return name" can be used by children nodes to propagate results.

GetCode sounds like translate, which this function is absolutely not.

Reply:
Yes okay I understand, but it does not return `"z"`, but it returns the expression on the RHS the assignment, no? And that's not a name

Reply:
Not necessarily. There are plenty of cases where you would return z (for example nodes that create loops!)

Reply:
Right, there are cases where the expression is equal to the name, but this doesn't appear to be the general case. Why calling a function `getResultName()` if it only returns the name in a subset of cases? Other alternatives are `getResult()` or `getExpression()`.